File: ft8af/app/src/main/kotlin/radio/ks3ckc/ft8af/wsjtx/WsjtxUdpService.kt
package radio.ks3ckc.ft8af.wsjtx

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.k1af.ft8af.GeneralVariables
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.io.File
import java.io.FileWriter
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.net.InetSocketAddress
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * WSJT-X UDP transport for Android: owns the socket, broadcasts outbound
 * datagrams built by [WsjtxCodec], and (when "accept requests" is on) runs a
 * background listener that turns inbound requests into callbacks the ViewModel
 * wires to the transmit path.
 *
 * Modelled on [radio.ks3ckc.ft8af.pskreporter.PskReporterSender] (an `object`
 * with an IO coroutine scope + `DatagramSocket`). Settings live in
 * [GeneralVariables] (`udpEnabled`/`udpHost`/`udpPort`/`udpAcceptRequests`);
 * [reload] (re)binds from them, so both startup and a Settings change route
 * through one place.
 *
 * Inbound callbacks are posted to the main thread, since they drive the same
 * transmit entry points as UI taps.
 */
object WsjtxUdpService {
    private const val TAG = "WsjtxUdpService"
    private const val REPLAY_CACHE_CAP = 1000
    private const val STATUS_INTERVAL_MS = 1000L
    private const val HEARTBEAT_INTERVAL_MS = 15_000L
    private const val RECV_TIMEOUT_MS = 500

    private val mainHandler = Handler(Looper.getMainLooper())

    private var scope: CoroutineScope? = null
    private var socket: DatagramSocket? = null
    private var target: InetAddress? = null
    private var targetPort = 0
    private var listenerThread: Thread? = null
    @Volatile private var listening = false
    // Guarded by [replayLock]: mutated from decode/band-change callers and read from the
    // listener thread handling a Replay request, so all access must be synchronized —
    // ArrayDeque is not thread-safe.
    private val replayCache = ArrayDeque<WsjtxCodec.Decode>()
    private val replayLock = Any()

    // Java-friendly SAM interfaces so the (Java) ViewModel can wire these with
    // plain lambdas.
    fun interface StatusProvider { fun get(): WsjtxCodec.Status? }
    fun interface ReplyHandler { fun onReply(call: String, grid: String, snr: Int) }
    fun interface HaltHandler { fun onHalt(autoOnly: Boolean) }
    fun interface FreeTextHandler { fun onFreeText(text: String, send: Boolean) }
    fun interface ReplayHandler { fun onReplay() }

    /** Supplies the current [WsjtxCodec.Status] snapshot for the periodic push. */
    @JvmField var statusProvider: StatusProvider? = null

    // Inbound request handlers (set by the ViewModel; invoked on the main thread).
    @JvmField var onReply: ReplyHandler? = null
    @JvmField var onReplay: ReplayHandler? = null
    @JvmField var onHaltTx: HaltHandler? = null
    @JvmField var onFreeText: FreeTextHandler? = null

    val enabled: Boolean get() = socket != null

    /** (Re)bind from the persisted [GeneralVariables] settings. Idempotent. */
    fun reload() {
        teardown()
        if (!GeneralVariables.udpEnabled) {
            log("disabled")
            return
        }
        try {
            target = InetAddress.getByName(GeneralVariables.udpHost)
            targetPort = GeneralVariables.udpPort
            // When accepting requests, bind the configured UDP port (default 2237) so
            // companion apps that send Reply/Halt/FreeText/Replay to that port actually
            // reach us. An ephemeral local port would only work for outbound broadcasts —
            // no companion could discover it. Send-only mode keeps the ephemeral port.
            val localPort = if (GeneralVariables.udpAcceptRequests) targetPort else 0
            val sock = DatagramSocket(null as java.net.SocketAddress?).apply {
                reuseAddress = true
                bind(InetSocketAddress(localPort))
                broadcast = true
            }
            socket = sock

            val s = CoroutineScope(Dispatchers.IO + SupervisorJob())
            scope = s
            s.launch { periodicLoop() }

            if (GeneralVariables.udpAcceptRequests) startListener(sock)
            log("bound → ${GeneralVariables.udpHost}:${targetPort}" +
                if (GeneralVariables.udpAcceptRequests) " (accepting requests)" else "")
        } catch (e: Exception) {
            log("bind failed: ${e.javaClass.simpleName}: ${e.message}")
            teardown()
        }
    }

    /** Tear down socket + loops. Safe to call when already stopped. */
    fun stop() = teardown()

    private fun teardown() {
        listening = false
        scope?.cancel()
        scope = null
        socket?.close()
        socket = null
        target = null
        // The listener's blocking receive unblocks via SO_TIMEOUT + the closed
        // socket; give it a moment to exit but don't hang teardown on it.
        listenerThread?.let { t -> t.join(1000) }
        listenerThread = null
    }

    private fun send(datagram: ByteArray) {
        val sock = socket ?: return
        val addr = target ?: return
        try {
            sock.send(DatagramPacket(datagram, datagram.size, addr, targetPort))
        } catch (e: Exception) {
            log("send error: ${e.javaClass.simpleName}: ${e.message}")
        }
    }

    // --- outbound -----------------------------------------------------------

    fun sendClear() {
        if (enabled) send(WsjtxCodec.clear())
    }

    fun sendDecode(d: WsjtxCodec.Decode) {
        if (!enabled) return
        send(WsjtxCodec.decode(d))
        synchronized(replayLock) {
            if (replayCache.size >= REPLAY_CACHE_CAP) replayCache.removeFirst()
            replayCache.addLast(d)
        }
    }

    fun sendQsoLogged(q: WsjtxCodec.QsoLogged, adif: String) {
        if (!enabled) return
        send(WsjtxCodec.qsoLogged(q))
        send(WsjtxCodec.loggedAdif(adif))
    }

    fun clearReplayCache() = synchronized(replayLock) { replayCache.clear() }

    /** Re-broadcast this session's decodes (marked not-new) for a Replay request. */
    private fun replay() {
        if (!enabled) return
        // Snapshot under the lock, then send outside it (send() must not hold replayLock).
        val snapshot = synchronized(replayLock) { replayCache.toList() }
        for (d in snapshot) send(WsjtxCodec.decode(d.copy(isNew = false)))
    }

    private suspend fun periodicLoop() {
        val s = scope ?: return
        var sinceHeartbeatMs = HEARTBEAT_INTERVAL_MS // send one immediately
        while (s.isActive) {
            statusProvider?.get()?.let { send(WsjtxCodec.status(it)) }
            if (sinceHeartbeatMs >= HEARTBEAT_INTERVAL_MS) {
                send(WsjtxCodec.heartbeat(GeneralVariables.VERSION, ""))
                sinceHeartbeatMs = 0
            }
            delay(STATUS_INTERVAL_MS)
            sinceHeartbeatMs += STATUS_INTERVAL_MS
        }
    }

    // --- inbound listener ---------------------------------------------------

    private fun startListener(sock: DatagramSocket) {
        sock.soTimeout = RECV_TIMEOUT_MS
        listening = true
        val t = Thread({
            val buf = ByteArray(2048)
            while (listening) {
                try {
                    val packet = DatagramPacket(buf, buf.size)
                    sock.receive(packet)
                    dispatch(packet.data.copyOf(packet.length))
                } catch (e: java.net.SocketTimeoutException) {
                    // loop to re-check `listening`
                } catch (e: Exception) {
                    if (listening) log("recv error: ${e.javaClass.simpleName}: ${e.message}")
                    if (sock.isClosed) break
                }
            }
        }, "ft8af-wsjtx-rx")
        t.isDaemon = true
        t.start()
        listenerThread = t
    }

    private fun dispatch(buf: ByteArray) {
        when (val msg = WsjtxCodec.parseInbound(buf)) {
            is WsjtxCodec.Inbound.Reply ->
                mainHandler.post { onReply?.onReply(msg.call, msg.grid, msg.snr) }
            is WsjtxCodec.Inbound.HaltTx ->
                mainHandler.post { onHaltTx?.onHalt(msg.autoOnly) }
            is WsjtxCodec.Inbound.FreeText ->
                mainHandler.post { onFreeText?.onFreeText(msg.text, msg.send) }
            WsjtxCodec.Inbound.Replay -> {
                replay()
                mainHandler.post { onReplay?.onReplay() }
            }
            null -> {} // ignored/unparseable
        }
    }

    private fun log(msg: String) {
        Log.d(TAG, msg)
        try {
            val ctx = GeneralVariables.getMainContext() ?: return
            val dir = ctx.getExternalFilesDir(null) ?: return
            val ts = SimpleDateFormat("HH:mm:ss.SSS", Locale.US).format(Date())
            FileWriter(File(dir, "debug.log"), true).use { it.append("$ts WsjtxUdpService: $msg\n") }
        } catch (_: Exception) {
        }
    }
}

File: ios/FT8AFKit/Sources/FT8Engine/WsjtxUdpService.swift
import Foundation
import Network

/// WSJT-X UDP transport for iOS: owns an outbound `NWConnection` for broadcasts,
/// an `NWListener` (when "accept requests" is on) bound to the configured port
/// for inbound requests, and turns those requests into callbacks the engine
/// wires to the transmit path.
///
/// Cadence (Status ~1 Hz, Heartbeat ~15 s) is driven by the caller (LiveEngine),
/// which already ticks per slot and holds the state a Status needs — so this
/// class stays a thin I/O shim. Inbound callbacks are delivered on the main
/// queue, matching the engine's `@MainActor` isolation.
///
/// Threading: all `connection`/`listener`/`replayCache`/`config` access is
/// confined to `queue` (a serial queue), so sends and the replay cache can't
/// race across the send callers and the listener's receive callbacks. A separate
/// lock-guarded flag backs the synchronous `isEnabled`.
public final class WsjtxUdpService {
    public struct Config: Equatable {
        public var enabled: Bool
        public var host: String
        public var port: UInt16
        public var acceptRequests: Bool
        public init(enabled: Bool = false, host: String = "127.0.0.1",
                    port: UInt16 = 2237, acceptRequests: Bool = false) {
            self.enabled = enabled; self.host = host; self.port = port
            self.acceptRequests = acceptRequests
        }
    }

    private let queue = DispatchQueue(label: "radio.ks3ckc.ft8af.wsjtx")
    // All of the following are touched only on `queue`.
    private var connection: NWConnection?
    private var listener: NWListener?
    private var config = Config()
    private var replayCache: [WsjtxCodec.Decode] = []
    private let replayCacheCap = 1000

    // isEnabled is read synchronously from the (main-actor) engine, so it is backed
    // by a lock-guarded flag set eagerly in apply()/stop() rather than by reading the
    // queue-confined `connection`.
    private let stateLock = NSLock()
    private var enabledFlag = false

    // Inbound request handlers, invoked on the main queue.
    public var onReply: ((_ call: String, _ grid: String, _ snr: Int32, _ deltaFreq: UInt32) -> Void)?
    public var onReplay: (() -> Void)?
    public var onHaltTx: ((_ autoOnly: Bool) -> Void)?
    public var onFreeText: ((_ text: String, _ send: Bool) -> Void)?

    public init() {}

    public var isEnabled: Bool {
        stateLock.lock(); defer { stateLock.unlock() }
        return enabledFlag
    }

    private func setEnabled(_ v: Bool) {
        stateLock.lock(); enabledFlag = v; stateLock.unlock()
    }

    /// (Re)configure: tears down any existing connection/listener and reconnects
    /// if enabled. Safe to call repeatedly (Settings changes route through here).
    public func apply(_ cfg: Config) {
        // Set the synchronous flag eagerly so an isEnabled check right after apply()
        // reflects intent, then do the socket work on the serial queue.
        let willEnable = cfg.enabled && NWEndpoint.Port(rawValue: cfg.port) != nil
        setEnabled(willEnable)
        queue.async { [weak self] in
            guard let self else { return }
            self.teardownLocked()
            self.config = cfg
            guard willEnable, let port = NWEndpoint.Port(rawValue: cfg.port) else { return }

            // Outbound connection for Status/Decode/Heartbeat/Clear broadcasts.
            let conn = NWConnection(host: NWEndpoint.Host(cfg.host), port: port, using: .udp)
            self.connection = conn
            conn.start(queue: self.queue)

            // Inbound requests need a listener BOUND to the configured port — companion
            // apps send Reply/Halt/FreeText/Replay there. The outbound connection's
            // ephemeral local port is not discoverable, so it can never receive them.
            if cfg.acceptRequests {
                self.startListenerLocked(port: port)
            }
        }
    }

    public func stop() {
        setEnabled(false)
        queue.async { [weak self] in self?.teardownLocked() }
    }

    /// Must run on `queue`.
    private func teardownLocked() {
        connection?.cancel(); connection = nil
        listener?.cancel(); listener = nil
    }

    // MARK: - outbound

    /// Build (on `queue`, since the codec is pure) and send a datagram. No-op when
    /// there is no connection.
    private func enqueueSend(_ build: @escaping () -> Data) {
        queue.async { [weak self] in
            guard let self, let conn = self.connection else { return }
            conn.send(content: build(), completion: .idempotent)
        }
    }

    public func sendClear() { enqueueSend { WsjtxCodec.clear() } }

    public func sendStatus(_ s: WsjtxCodec.Status) { enqueueSend { WsjtxCodec.status(s) } }

    public func sendHeartbeat(version: String) {
        enqueueSend { WsjtxCodec.heartbeat(version: version, revision: "") }
    }

    public func sendDecode(_ d: WsjtxCodec.Decode) {
        queue.async { [weak self] in
            guard let self, let conn = self.connection else { return }
            conn.send(content: WsjtxCodec.decode(d), completion: .idempotent)
            if self.replayCache.count >= self.replayCacheCap { self.replayCache.removeFirst() }
            self.replayCache.append(d)
        }
    }

    public func sendQsoLogged(_ q: WsjtxCodec.QsoLogged, adif: String) {
        enqueueSend { WsjtxCodec.qsoLogged(q) }
        enqueueSend { WsjtxCodec.loggedAdif(adif) }
    }

    public func clearReplayCache() {
        queue.async { [weak self] in self?.replayCache.removeAll() }
    }

    /// Re-broadcast this session's decodes (marked not-new) for a Replay request.
    /// Must run on `queue`.
    private func replay() {
        guard let conn = connection else { return }
        for var d in replayCache {
            d.isNew = false
            conn.send(content: WsjtxCodec.decode(d), completion: .idempotent)
        }
    }

    // MARK: - inbound

    /// Must run on `queue`.
    private func startListenerLocked(port: NWEndpoint.Port) {
        do {
            let params = NWParameters.udp
            params.allowLocalEndpointReuse = true
            let l = try NWListener(using: params, on: port)
            l.newConnectionHandler = { [weak self] conn in
                guard let self else { conn.cancel(); return }
                conn.start(queue: self.queue)
                self.receive(on: conn)
            }
            l.start(queue: queue)
            listener = l
        } catch {
            // Port unavailable: inbound won't work, but outbound broadcasts still do.
            listener = nil
        }
    }

    /// Receive callbacks fire on `queue` (the connection was started on it), so
    /// `handleInbound` — including `replay()` — runs on `queue`.
    private func receive(on conn: NWConnection) {
        conn.receiveMessage { [weak self] data, _, _, error in
            guard let self else { return }
            if let data, !data.isEmpty { self.handleInbound(data) }
            if error == nil { self.receive(on: conn) }
        }
    }

    /// Must run on `queue`.
    private func handleInbound(_ data: Data) {
        guard let msg = WsjtxCodec.parseInbound(data) else { return }
        switch msg {
        case let .reply(call, grid, snr, deltaFreq):
            DispatchQueue.main.async { self.onReply?(call, grid, snr, deltaFreq) }
        case let .haltTx(autoOnly):
            DispatchQueue.main.async { self.onHaltTx?(autoOnly) }
        case let .freeText(text, send):
            DispatchQueue.main.async { self.onFreeText?(text, send) }
        case .replay:
            replay()
            DispatchQueue.main.async { self.onReplay?() }
        }
    }
}
